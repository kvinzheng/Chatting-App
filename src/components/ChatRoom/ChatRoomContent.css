.ChatRoomContent {
  width: 100%;
  height: 100%;

  display: flex;
  flex-direction: column;
  background-color: #e0e0e0;
}
.ChatRoomContent-messages {
  width: 100%;
  height: 100%;

  background-color: #e0e0e0;
  display: flex;
  flex-direction: column;
  overflow: scroll;
}

.ChatRoomContent-inputBoxWrapper {
  background-color: white;
  padding: 25px 35px;
  display: flex;
}

.ChatRoomContent-submitButton {
  background-color: white;
  border: none;
  color: #6495ed;
  font-size: 17px;
}
.ChatRoomContent-submitButton:hover {
  cursor: pointer;
}
