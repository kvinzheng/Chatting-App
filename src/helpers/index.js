import { ONE_MINUTE } from "../consts";

export const getOnlineMinute = (startTime, currentTime) => {
  const timeInterval = currentTime - startTime;
  return Math.ceil(timeInterval / ONE_MINUTE);
};
