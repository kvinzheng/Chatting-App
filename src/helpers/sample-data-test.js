export const ChatRoomContentSampleData = {
  messagesDetail: {
    0: [
      {
        name: "Kevin",
        message: "ayyyyy",
        id: "gg35545",
      },
      {
        name: "Nicole",
        message: "lmao",
        id: "yy35578",
      },
      {
        name: "Thomas",
        message: "leggooooo",
        id: "hh9843",
      },
      {
        name: "kevin",
        message: "kevin here",
        id: "ZdmAdcv7om12eiy8B9GUv",
      },
      {
        name: "kevin",
        message: "I am typing something here",
        id: "dtpL_v38S3l0REbU4Bp1K",
      },
      {
        name: "kevin",
        message: "hi",
        id: "ergyguhk-u6Y5dMqLuB7V",
      },
    ],
  },
  activeRoom: {
    name: "Art History Chat",
    id: 0,
    users: ["Kevin", "Nicole", "Thomas"],
  },
};
