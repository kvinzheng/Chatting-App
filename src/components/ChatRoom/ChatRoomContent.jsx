/* eslint-disable react-hooks/exhaustive-deps */
import React, { useEffect, useState, useRef } from "react";
import { connect } from "react-redux";
import socketIOClient from "socket.io-client";
import PropTypes from "prop-types";

import InputBox from "../../common/InputBox";
import Message from "../../common/Message";
import ChatRoomContentHeader from "./ChatRoomContentHeader";
import { fetchSingleRoom, updateActiveUsers } from "../../actions/room";
import {
  postMessageSingleRoom,
  fetchSingleRoomMessages,
} from "../../actions/message";

import "./ChatRoomContent.css";

import { SCROLL_OPTIONS, NEW_CHAT_MESSAGE, SERVER_HOST } from "../../consts";

export const ChatRoomContent = ({
  currentUserName,
  postMessageSingleRoom,
  fetchSingleRoom,
  fetchSingleRoomMessages,
  messagesDetail,
  activeRoom,
  updateActiveUsers,
}) => {
  const [socket, setSocket] = useState(null);
  const [inputValue, setInputValue] = useState("");
  const inputBoxRef = useRef();
  const activeRoomId = activeRoom.id || 0;
  const messagesList = messagesDetail[activeRoomId] || [];

  const createSocketIO = () => {
    const socket = socketIOClient(SERVER_HOST, {
      query: { roomId: activeRoomId, user: currentUserName },
    });
    socket.on(NEW_CHAT_MESSAGE, ({ message, room }) => {
      const newactiveUsers = room.users;
      const validUserCheck =
        newactiveUsers &&
        activeRoom.users &&
        newactiveUsers.join("/") !== activeRoom.users.join("/");

      if (validUserCheck) {
        updateActiveUsers(newactiveUsers);
      }
      postMessageSingleRoom(activeRoomId, message);
    });
    return socket;
  };

  //create socket when id change
  useEffect(() => {
    const fetApi = async () => {
      await Promise.all([
        fetchSingleRoom(activeRoomId),
        fetchSingleRoomMessages(activeRoomId),
      ]);
    };
    fetApi();

    const newSocket = createSocketIO();
    setSocket(newSocket);
    return () => newSocket.disconnect();
  }, [activeRoomId]);

  //scroll to the inputbox
  useEffect(() => {
    inputBoxRef.current.scrollIntoView(SCROLL_OPTIONS);
  }, [messagesList.length]);

  const handleMessageSubmit = (e) => {
    e.preventDefault();
    socket.emit(NEW_CHAT_MESSAGE, inputValue);
    setInputValue("");
  };

  return (
    <div className="ChatRoomContent">
      <ChatRoomContentHeader
        currentUserName={currentUserName}
        userNamesList={activeRoom.users}
        name={activeRoom.name}
      />

      <div className="ChatRoomContent-messages">
        {messagesList.map(({ name, message }, index) => {
          return (
            <Message
              myself={name === currentUserName}
              author={name}
              key={index}
            >
              {message}
            </Message>
          );
        })}
        <div className="ChatRoomScroll" ref={inputBoxRef} />
      </div>
      <form
        className="ChatRoomContent-inputBoxWrapper"
        onSubmit={handleMessageSubmit}
      >
        <InputBox
          placeholder="Type a message..."
          name="message-input"
          value={inputValue}
          onChange={(e) => setInputValue(e.target.value)}
        />
        <button
          type="link"
          className="ChatRoomContent-submitButton"
          onClick={handleMessageSubmit}
        >
          Send
        </button>
      </form>
    </div>
  );
};
export const mapStateToProps = (state) => {
  return {
    messagesDetail: state.messages,
    rooms: state.rooms.data,
    activeRoom: state.rooms.active,
  };
};

ChatRoomContent.propTypes = {
  messagesDetail: PropTypes.object,
  rooms: PropTypes.array,
  activeRoom: PropTypes.object,
  updateActiveUsers: PropTypes.func,
};
export default connect(mapStateToProps, {
  postMessageSingleRoom,
  fetchSingleRoom,
  fetchSingleRoomMessages,
  updateActiveUsers,
})(ChatRoomContent);
