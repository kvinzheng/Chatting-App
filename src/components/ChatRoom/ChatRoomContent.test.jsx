import React from "react";
import { shallow, configure, mount } from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

import { ChatRoomContent } from "./ChatRoomContent.jsx";
import { ChatRoomContentSampleData } from "../../helpers/sample-data-test";

configure({ adapter: new Adapter() });

const postMessageSingleRoom = jest.fn();
const fetchSingleRoom = jest.fn();
const fetchSingleRoomMessages = jest.fn();
const updateActiveUsers = jest.fn();

describe("ChatRoomContent Component", () => {
  beforeEach(() => {});
  it("should render with testing data ", () => {
    const component = shallow(
      <ChatRoomContent
        currentUserName={"Nick"}
        postMessageSingleRoom={postMessageSingleRoom}
        fetchSingleRoom={fetchSingleRoom}
        fetchSingleRoomMessages={fetchSingleRoomMessages}
        messagesDetail={{ 0: [] }}
        activeRoom={{ id: 0, name: "", users: [] }}
        updateActiveUsers={updateActiveUsers}
      />
    );
    expect(component).toMatchSnapshot();
  });

  it("Renders with correct classes", () => {
    const component = shallow(
      <ChatRoomContent
        currentUserName={"Nick"}
        postMessageSingleRoom={postMessageSingleRoom}
        fetchSingleRoom={fetchSingleRoom}
        fetchSingleRoomMessages={fetchSingleRoomMessages}
        messagesDetail={{ 0: [] }}
        activeRoom={{ id: 0, name: "", users: [] }}
        updateActiveUsers={updateActiveUsers}
      />
    );
    expect(component.find(".ChatRoomContent-messages").length).toBe(1);
    expect(component.find(".ChatRoomContent-inputBoxWrapper").length).toBe(1);
    expect(component.find(".ChatRoomContent-submitButton").length).toBe(1);
    expect(component.find(".ChatRoomContent").length).toBe(1);
  });

  it("functional test: simulate a input text and then expect the text show up on the chat room while scrolling down to the bottom of the chat", () => {
    const scrollIntoView = jest.fn();
    window.HTMLElement.prototype.scrollIntoView = scrollIntoView;
    const inputBoxRef = jest.spyOn(React, "useRef");
    inputBoxRef.current = { scrollIntoView: scrollIntoView };

    const wrapper = mount(
      <ChatRoomContent
        currentUserName={"Nick"}
        postMessageSingleRoom={postMessageSingleRoom}
        fetchSingleRoom={fetchSingleRoom}
        fetchSingleRoomMessages={fetchSingleRoomMessages}
        messagesDetail={ChatRoomContentSampleData.messagesDetail}
        activeRoom={ChatRoomContentSampleData.activeRoom}
        updateActiveUsers={updateActiveUsers}
      />
    );

    const input = wrapper.find("InputBox");
    const button = wrapper.find(".ChatRoomContent-submitButton");

    input.simulate("change", {
      target: { value: "I am typing something here" },
    });
    button.simulate("click");

    expect(inputBoxRef.current.scrollIntoView).toHaveBeenCalledTimes(1);
    expect(wrapper.find(".ChatRoomContent-messages").length).toBe(1);
    expect(wrapper.find(".Message-body").length).toBe(6);
    expect(wrapper.text().includes("I am typing something here")).toBe(true);
  });
});
