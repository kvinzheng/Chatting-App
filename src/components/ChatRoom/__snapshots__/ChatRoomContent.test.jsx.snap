// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatRoomContent Component should render with testing data  1`] = `ShallowWrapper {}`;
